src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1104,10 +1104,75 @@ private LLVMValueRef CastToPointerToTypeDesc(LLVMValueRef source, TypeDesc type,
+                {
+                    StoreStruct(address, llvmValue, targetType, typedStoreLocation);
+                }
+                else
+                {
+                    _builder.BuildStore(llvmValue, typedStoreLocation);
+                }
+            }
+        }
+
+        private static bool IsStruct(TypeDesc typeDesc)
+        {
+            return typeDesc.IsValueType && !typeDesc.IsPrimitive && !typeDesc.IsEnum;
+        }
+
+        private void StoreStruct(LLVMValueRef address, LLVMValueRef llvmValue, TypeDesc targetType, LLVMValueRef typedStoreLocation, bool childStruct = false)
+        {
+            foreach (FieldDesc f in targetType.GetFields())
+            {
+                if (f.IsStatic) continue;
+                if (IsStruct(f.FieldType) && llvmValue.TypeOf.IsPackedStruct)
+                {
+                    LLVMValueRef targetAddress = _builder.BuildGEP(address, new[] { BuildConstInt32(f.Offset.AsInt) });
+                    uint index = LLVMSharpInterop.ElementAtOffset(_compilation.TargetData, llvmValue.TypeOf, (ulong)f.Offset.AsInt);
+                    LLVMValueRef fieldValue = _builder.BuildExtractValue(llvmValue, index);
+                    //recurse into struct
+                    StoreStruct(targetAddress, fieldValue, f.FieldType, CastToPointerToTypeDesc(targetAddress, f.FieldType), true);
+                }
+                else if (f.FieldType.IsGCPointer)
+                {
+                    LLVMValueRef targetAddress = _builder.BuildGEP(address, new[] {BuildConstInt32(f.Offset.AsInt)});
+                    LLVMValueRef fieldValue;
+                    if (llvmValue.TypeOf.IsPackedStruct)
+                    {
+                        uint index = LLVMSharpInterop.ElementAtOffset(_compilation.TargetData, llvmValue.TypeOf, (ulong) f.Offset.AsInt);
+                        fieldValue = _builder.BuildExtractValue(llvmValue, index);
+                        Debug.Assert(fieldValue.TypeOf.Kind == LLVMTypeKind.LLVMPointerTypeKind, "expected an LLVM pointer type");
+                    }
+                    else
+                    {
+                        // single field IL structs are not LLVM structs
+                        fieldValue = llvmValue;
+                    }
+                    CallRuntime(_method.Context, "InternalCalls", "RhpAssignRef",
+                        new StackEntry[]
+                        {
+                            new ExpressionEntry(StackValueKind.Int32, "targetAddress", targetAddress),
+                            new ExpressionEntry(StackValueKind.ObjRef, "sourceAddress", fieldValue)
+                        });
+                }
+            }
+            if (!childStruct)
+            {
+                _builder.BuildStore(llvmValue, typedStoreLocation); // just copy all the fields again for simplicity, if all the fields where set using RhpAssignRef then a possible optimisation would be to skip this line
```

Comment:
This has a potential for race conditions if the LLVM codegen is used for anything multithreaded. Maybe add a comment about it?

Reply:
Right, think I understand - thread1 can RhpAssignRef a struct field (after thread2 has done the same), thread1 calls `BuildStore` and then thread2 can get switched back in and executes `BuildStore`.   Thread 2's field which requires the GC write barrier is then lost.  Will add a comment for now.